pkg/generator/gc/gc.go

```diff
@@ -0,0 +1,90 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package gc
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+	"time"
+
+	"github.com/go-co-op/gocron/v2"
+	"github.com/spf13/pflag"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	genv1alpha1 "github.com/external-secrets/external-secrets/apis/generators/v1alpha1"
+	"github.com/external-secrets/external-secrets/pkg/feature"
+)
+
+var gcGracePeriod time.Duration
+var log = ctrl.Log.WithName("generator-gc")
+var scheduler gocron.Scheduler
+
+func init() {
+	fs := pflag.NewFlagSet("gc", pflag.ExitOnError)
+	fs.DurationVar(&gcGracePeriod, "generator-gc-grace-period", time.Minute*15, "Duration after which generated secrets are cleaned up after they have been flagged for gc.")
+	feature.Register(feature.Feature{
+		Flags: fs,
+	})
+	var err error
+	scheduler, err = gocron.NewScheduler()
+	if err != nil {
+		panic(err)
+	}
+	scheduler.Start()
+}
+
+type Entry struct {
+	Resource *apiextensions.JSON
+	Impl     genv1alpha1.Generator
+	State    genv1alpha1.GeneratorProviderState
+}
+
+func (e Entry) Key() string {
+	h := sha256.New()
+	h.Write(e.Resource.Raw)
+	hash := h.Sum(e.State.Raw)
+	return fmt.Sprintf("%x", hash)
+}
+
+func Enqueue(e Entry) error {
+	log.V(1).Info("putting state into GC", "entry", e)
+	_, err := scheduler.NewJob(
```

Comment:
Should we care here for things like:
- jobs for entires that already have pending jobs
- cleaning up jobs that got stuck
- schedule not running because of pod restart

Reply:
> schedule not running because of pod restart

The schedule is being lost after pod restart, that's right. But it's not that important. `GarbageCollect()` is called after each reconciliation cycle, that's when it eventually gets cleaned up. So worst case it should be cleaned up after `spec.refreshInterval`. We can even consider dropping the schedule all together and solely rely on `GarbageCollect()` being called from a reconcile loop.
The `.Cleanup()` is idempotent, so we can safely call it from multiple places. Either from the schedule, when `GarbageCollect()` is called or the resource is being deleted.

> cleaning up jobs that got stuck

Not sure how/if jobs gets stuck?

> jobs for entries that already have pending jobs

hmm, `Enqueue()` is called only on a state transition, when moving `latest -> gc`. calling `.Cleanup()` multiple times is not a problem because it is idempotent.

---
maybe - just for simplicity - should i consider dropping the whole schedule bit? It's easier to reason about the gc process when there is no delayed/scheduled cleanup and everything is processed inside the `Reconcile()` loop :thinking: 

The schedule is just a optimization to delete the state independent of the reconcile loop, so we don't have to rely on it.

Reply:
It's fine, I like it _because_ it doesn't burden the reconcile loop. It's already burdened enough. :D